vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/dao/impl/DeviceRepositoryImpl.java

```diff
@@ -0,0 +1,175 @@
+        return CACHE_DEVICE_INFO.get(uid);
+    }
+
+    @Override
+    public Storage getLoginedDeviceByIP(String deviceIP) {
+        return LOGINED_DEVICE.get(deviceIP);
+    }
+
+
+    private void writeDataToFile() {
+        ObjectOutputStream objectOutputStream = null;
+        try {
+            File file = new File(DEVICE_DATA_FILE_PATH);
+            objectOutputStream = new ObjectOutputStream(new FileOutputStream(file));
+            synchronized (CACHE_DEVICE_INFO) {
+                objectOutputStream.writeObject(CACHE_DEVICE_INFO);
+            }
+        } catch (Exception e) {
+            logger.error(e.getMessage(),e);
+        } finally {
+            quietClose(objectOutputStream);
+        }
+    }
+
+    private void initDevice(DeviceInfo deviceInfo) throws Exception {
+        Storage device = StorageFactory.newStorage(getDeviceClassName(deviceInfo.deviceType), "");
+        device.login(deviceInfo.ip, deviceInfo.port, deviceInfo.username, deviceInfo.password);
+        StorageMO deviceMO = device.getDeviceInfo();
+        deviceInfo.deviceModel = deviceMO.model;
+        deviceInfo.deviceName = deviceMO.name;
+        deviceInfo.deviceStatus = deviceMO.status;
+        deviceInfo.sn = deviceMO.sn;
+        LOGINED_DEVICE.put(deviceInfo.ip, device);
+    }
+
+    private String getDeviceClassName(String deviceType) {
+        String[] types = StorageFactory.listStorages();
+        for (String type : types) {
+            if (type.substring(type.lastIndexOf(".") + 1).equals(deviceType)) {
+                return type;
+            }
+        }
+        return "";
+    }
+
+    private void quietClose(Closeable closeable) {
+        if (closeable != null) {
+            try {
+                closeable.close();
+            } catch (IOException e) {
+                //ignore it
+            }
+        }
+    }
+
+    private void reLogin(Map<String, DeviceInfo> deviceInfoMap) {
+        if (deviceInfoMap == null || deviceInfoMap.isEmpty()) {
+            return;
+        }
+        ExecutorService executorService = Executors.newCachedThreadPool();
+        int deviceNumber = deviceInfoMap.size();
+        final CountDownLatch countDownLatch = new CountDownLatch(deviceNumber);
+        for (final Map.Entry<String, DeviceInfo> entry : deviceInfoMap.entrySet()) {
+            executorService.submit(new Runnable() {
+                @Override
+                public void run() {
+                    try {
+                        DeviceRepositoryImpl.this.initDevice(entry.getValue());
+                    } catch (Exception e) {
+                        logger.error("The device " + entry.getValue().ip + " fail to login");
+                    } finally {
+                        countDownLatch.countDown();
+                    }
+                }
+            });
+        }
+        try {
+            countDownLatch.await(5, TimeUnit.MINUTES);
+        } catch (InterruptedException e) {
+
```

Comment:
Why nothing in this catch block?

Reply:
Thanks，we will add the error logger here.

Reply:
Done